### PR TITLE
Freefem with petsc

### PR DESCRIPTION
**Pull request by corentin-dev**
Adding new software:
- tetgen
- parmmg
- hpddm
- freefem

### CODE REVIEW

File: var/spack/repos/builtin/packages/hpddm/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/hpddm/package.py]: These comments need to be removed before merging

Comment (corentin-dev) [reply var/spack/repos/builtin/packages/hpddm/package.py]: I removed those comments, and also improved the style of some `package.py`. Thank you for your remark.

**Comment by corentin-dev** [comment]
Dear @alalazo , the change I've made should do the work, can you confirm ?
Thank you

File: var/spack/repos/builtin/packages/petsc/package.py
Comment (balay) [inline var/spack/repos/builtin/packages/petsc/package.py]: Ah - circular dependency ...

File: var/spack/repos/builtin/packages/petsc/package.py
Comment (balay) [inline var/spack/repos/builtin/packages/petsc/package.py]: Presumably this code can be re-formatted wrt changes in https://github.com/spack/spack/pull/25074 (so rebase after that one is merged?)

Comment (corentin-dev) [reply var/spack/repos/builtin/packages/petsc/package.py]: Very well, I'll do so.

File: var/spack/repos/builtin/packages/petsc/package.py
Comment (corentin-dev) [reply var/spack/repos/builtin/packages/petsc/package.py]: Actually, it seems to be a problem related to the way the dependency is imported into PETSc, I'll talk in more details with the developer.

Comment (balay) [reply var/spack/repos/builtin/packages/petsc/package.py]: Yeah - its unavoidable..

**Comment by alalazo** [comment]
@corentin-dev Style changes are fine with me. Ping me when the review is done with maintainers of the various packages touched here. I'll proceed merging the PR once they approve it.

File: var/spack/repos/builtin/packages/petsc/package.py
Comment (balay) [inline var/spack/repos/builtin/packages/petsc/package.py]: Perhaps:
```
depends_on('gmake', type='build')
```

File: var/spack/repos/builtin/packages/petsc/package.py
Comment (balay) [reply var/spack/repos/builtin/packages/petsc/package.py]: #25074 is merged - and this can be rebsed. Here  - the code here would be: (if --with-pkg-include,--with-pkg-lib would work - its preferable)
```patch
diff --git a/var/spack/repos/builtin/packages/petsc/package.py b/var/spack/repos/builtin/packages/petsc/package.py
index 76d93ac424..0461e0e5da 100644
--- a/var/spack/repos/builtin/packages/petsc/package.py
+++ b/var/spack/repos/builtin/packages/petsc/package.py
@@ -417,6 +417,9 @@ def install(self, spec, prefix):
                 (jpeg_sp, 'libjpeg', True, True),
                 (scalapack_sp, 'scalapack', False, True),
                 'strumpack',
+                'mmg',
+                'parmmg',
+                'tetgen',
         ):
             # Cannot check `library in spec` because of transitive deps
             # Cannot check variants because parmetis keys on +metis
```


However - if only --with-pkg-dir works and not the above - use the format:

(`mmg, 'mmg', False, False),`

Comment (corentin-dev) [reply var/spack/repos/builtin/packages/petsc/package.py]: I'm back, and changed the commits accordingly. I hope it fits the philosophy.
Tetgen failed with `'tetgen'` so I used `('tetgen', 'tetgen', False, False),`.

File: var/spack/repos/builtin/packages/petsc/package.py
Comment (corentin-dev) [reply var/spack/repos/builtin/packages/petsc/package.py]: Done! Thank you for the suggestion!

File: var/spack/repos/builtin/packages/petsc/package.py
Comment (balay) [inline var/spack/repos/builtin/packages/petsc/package.py]: PETSc build requires gnumake - so `when='+suite-sparse'` should be removed.

File: var/spack/repos/builtin/packages/petsc/package.py
Comment (balay) [reply var/spack/repos/builtin/packages/petsc/package.py]: Perhaps adding `include`, `libs` properties to tetgen/package.py  is a better fix?

File: var/spack/repos/builtin/packages/petsc/package.py
Comment (corentin-dev) [reply var/spack/repos/builtin/packages/petsc/package.py]: Ok I'll leave it like that.
It was a guess from:
`config/BuildSystem/config/packages/SuiteSparse.py:46:      raise RuntimeError('SuiteSparse buildtools require GNUMake. Use --with-make=gmake or --download-make')`
But probably other packages needs it.

Comment (balay) [reply var/spack/repos/builtin/packages/petsc/package.py]: Ah - ok. This check might have been added to petsc configure before petsc switched to requiring gnumake [so we could remove this check for suite-sparse from petsc..]

And that check is for `--download-suitesparse` - which should never get used when building via spack..

Also we haven't seen a failure in petsc build so far without this dependency - so  perhaps spack already defaults to using gnumake?

Comment (corentin-dev) [reply var/spack/repos/builtin/packages/petsc/package.py]: > Also we haven't seen a failure in petsc build so far without this dependency - so perhaps spack already defaults to using gnumake?

Actually, I added the dependency based on a failure. Though I now wonder if it was a problem with `gmake` or something else since `gmake` is checked system wide and CentOS `make` comes with `gmake`... Anyway, I am currently rebuilding `PETSc` with a bare CentOS linux without this dependency, I'll tell you the result as soon as I get it.

Comment (corentin-dev) [reply var/spack/repos/builtin/packages/petsc/package.py]: It works without `gmake` (branch `develop`), so I removed it.

**Comment by corentin-dev** [comment]
@alalazo @balay If you agree, I'm done with changes on this branch, and you can proceed to a merge.
Thanks a lot for your reviews.

File: var/spack/repos/builtin/packages/freefem/package.py
Comment (balay) [inline var/spack/repos/builtin/packages/freefem/package.py]: Typo: SLEPSc  -> SLEPc

Comment (corentin-dev) [reply var/spack/repos/builtin/packages/freefem/package.py]: Thank you, I just pushed a new commit.

**Review by balay — DISMISSED** [review]

File: var/spack/repos/builtin/packages/freefem/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/freefem/package.py:40]: Can you please add a brief comment above these patches explaining what they do, why they're needed and where they are taken from?

**Review by alalazo — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/freefem/package.py
Comment (corentin-dev) [reply var/spack/repos/builtin/packages/freefem/package.py:40]: I added a comment about it (configure not finding correct flags when using compilers with full path! Thank you

**Review by alalazo — DISMISSED** [review]

**Review by alalazo — APPROVED** [review]

File: var/spack/repos/builtin/packages/petsc/package.py
Comment (lrtfm) [inline var/spack/repos/builtin/packages/petsc/package.py:243]: When `+parmmg`, `+mmg` should be added for PETSc.

File: var/spack/repos/builtin/packages/parmmg/package.py
Comment (lrtfm) [inline var/spack/repos/builtin/packages/parmmg/package.py:32]: If we do not set the directory of `mmg`, `parmmg` will download `mmg` and then build it. Other than dependent on the `mmg` built by spack. Here, we should add `cmake` flags for `parmmg`, like this. But for now, this will fail for `parmmg` needs some header files from the source of `mmg`. Some patch of `mmg` is needed to make this work.
```
        define = CMakePackage.define
        args = [
            self.define_from_variant('CMAKE_POSITION_INDEPENDENT_CODE', 'pic'),
            define('DOWNLOAD_MMG', False),
            define('MMG_DIR', self.spec['mmg'].prefix)
        ]
```

A workaround for those problems are given in #31838.

Comment (lrtfm) [reply var/spack/repos/builtin/packages/parmmg/package.py:32]: A workaround for those problems are given in #31838.

**Pull request merged**
